### ISSUE TITLE
The New Editor

### ISSUE THREAD

**Issue by atesh** [atesh/XKit#195]
Hey everyone, I just got the new post editor. 

Seems like they've moved away from TinyMCE, which is bad news since now we need to decrypt obfuscated code. They also appear to have removed the ?tags URI parameter.

Anyone had a chance to mess with the code? Any information to speed this up since I'm already 2 days behind would be much appreciated. Thanks!

**Comment by gengkev** [comment]
I think there's some Backbone.js in there, and the editor code seems to be in https://secure.assets.tumblr.com/client/prod/app/context/dashboard/index.js?_v=9dd991a20a4a6544496fe79f0cea11a4 with references to things like .editor[contenteditable], tagEditor, and even handleTagInput, which pass around events, but it's not really clear how to get to those. Maybe that's not the best approach. Good luck with getting tags to work!

**Comment by atesh** [comment]
Yep, appears to be using Backbone heavily. I couldn't do much yesterday about it but except for the body part of the post, everything is stored in objects (body contents are read from DOM) so a simple "change the contents of this element to this" does not affect the post at all.

Most of the stuff seems to be stored in Tumblr.Prima object.

There is a post model, and a new view is created each time you want to create a new post / edit one. If you set a breakpoint to any event handlers and use "this.AddTag("[tag]")" it works great - but not knowing much about Backbone I have no idea how to get the current editor. 

Everything seems to be self-contained for each 'post editor' so I am not sure how we can get it.

**Comment by WyohKnott** [comment]
> They also appear to have removed the ?tags URI parameter.

The code is still there but not working.
```javascript
get_parse_params: function(d) {
            var e = window.location.pathname.split("/");
            e.shift();
            if (e[0] === "tagged") {
                return "/new/" + d + "?post[tags]=" + e[1]
            } else {
                return "/new/" + d
            }
},
```
Search a tag and open "Create a new post" in a new tab. 
For example with xkit we get: https://www.tumblr.com/new/text?post[tags]=xkit. But nothing shows up in the tag area of the new post.
It might be in our interest to report a bug.

**Comment by inklesspen** [comment]
I did a little poking around in the source. It's possible to access the backbone classes of the individual bits of the editor. For instance:

```js
var tagEditor = require(200);  // This constant is assigned to the tag editor in the compressed JS
var origAddTag = tagEditor.prototype.addTag;
tagEditor.prototype.addTag = function(t) {console.log(t); return origAddTag.apply(this, [t]);}
```

Once I execute this in Firefox console, adding any tags to a post causes them to be logged to the console too. I don't know exactly what functionality we need to access on the editor; if anyone wants to list that out, I can figure out how to access it.

**Comment by gengkev** [comment]
@inklesspen hey! that's a really cool idea. i was looking at too and was wondering how to get the correct instance, because I think tags need to be programmatically added (?). but if you run something like what you wrote on page load, for example

```javascript
var tagEditorInstance;
(function(){
    var tagEditor = require(200);
    var origInitialize = tagEditor.prototype.initialize;
    tagEditor.prototype.initialize = function() {
        tagEditorInstance = this;
        origInitialize.apply(this, arguments);
    };
}());
```

...then you can just use that instance to add tags.

**Comment by inklesspen** [comment]
Ok, so I've spent some fun time in the debugger, and here's the scoop.

On the dashboard page, there's a script tag containing the following:
```js
require("context").bootloader({}) // with a lot of bootstrap data in the call to bootloader
```

The return value of bootloader is the app object, but it's discarded. If we had some way of capturing that, we'd be in.

I'm gonna keep poking at this. I have a few more ideas.

**Comment by inklesspen** [comment]
Okay, here's something for you.

```js
var postFormClass = require(58);
var oldInit = postFormClass.prototype.initialize;
postFormClass.prototype.initialize = function(t) {window.postFormController = this; return oldInit.apply(this, arguments);}
```

Now, whenever the new post button is clicked, we get a reference to the postFormController stored on the window object. (We can trigger a click on the new post button through JS, which calls createPostForm on the PostFormBuilder; sadly the createPostForm function cannot be intercepted with this method; it's already bound to the event.)

postFormController has some useful properties.

```js
// hide the #tags placeholder (not done automatically via addTag
postFormController.postForm.tagEditor.tagInput.editor.setPlaceholder(false);
// show it again
postFormController.postForm.tagEditor.tagInput.editor.setPlaceholder(postFormController.postForm.tagEditor.placeholder);
// add a tag
postFormController.postForm.tagEditor.addTag("fnord")
```

**Comment by atesh** [comment]
Ah thank you guys! it's been a rough week and I could only work on this after midnight for the last two days so I'm stumped :p

@inklesspen, I've tried adding the last code to xkit_patches, but initialize never gets called - most likely due to it being called before xkit_patches is loaded?

maybe we can modify the prototype of tagEditor ("200") to make it set itself to a reference stored on the window object whenever one of it's functions are called?

**Comment by inklesspen** [comment]
@atesh sorry, my mistake; 58 is the constant for the PostFormBuilder (which as I mentioned we can't monkeypatch very well). The correct constant is 74.

I haven't tried doing it from an xkit extension, but when I execute the three lines from that first code block (with the correct constant) in the console and click "new text post", the postFormController exists.

**Comment by inklesspen** [comment]
That said, when I tried accessing the `require` function from an xkit extension in Chrome, I was unable to get it.

**Comment by atesh** [comment]
ah! now it works! and you need to use XKit.tools.add_function to let them run on the window. 
it works when I do this:

```js
XKit.tools.add_function( function() { 
	
	try { 
		window.postFormClass = require(74);
		var oldInit = window.postFormClass.prototype.initialize;
		window.postFormClass.prototype.initialize = function(t) {window.postFormController = this; return oldInit.apply(this, arguments);}
		console.log("Changed postFormClass prototype.");
	} catch(e) {
		console.log("Error: " + e.message);	
	}
	
}, true, "");

**Comment by atesh** [comment]
it's almost 4 am now so I should be heading to bed, but I'll mess around with it tomorrow to see what else we can do with this. 

as far as I know, the following doesn't work anymore:

* **One-Click Reply** - the tagging does not work
* **AutoTagger**
* **QuickTags** - I've the 'post listener' part to add buttons to the editor, but still no tag modifying, so it's useless
* **Reblog As Link** - this will take some reverse engineering too. The old code had "change_reblog_type" and I bet the new code has something like that too, but not sure where it is hidden now.

Once again, thank you all for all your help. I'll keep you updated as I make progress tomorrow.

**Comment by inklesspen** [comment]
Once you figure out how to get access to the postFormController from inside One-Click Reply's fill_post method, it should be easy to add the tags there. (I fiddled with it briefly but couldn't figure out the correct incantation to access it.)

**Comment by WyohKnott** [comment]
The only way to access these functions is to execute them with add_function because they won't work in the sandboxed environment. It will work if we can redefine the class before they are initialised like postFormClass, but not for the ones already initialised before Xkit is loaded. That includes setPostType.

**Comment by WyohKnott** [comment]
> Reblog As Link - this will take some reverse engineering too. The old code had "change_reblog_type" and I bet the new code has something like that too, but not sure where it is hidden now.

We could simulate user actions instead, with ```$(".reblog-as-type").click();$("[data-reblog-as-type='text']").click();```

**Comment by BenWiederhake** [comment]
"(any extension I'm missing?)" -> Yep, there's also ViewMyTags.

I'm not sure whether this is #192

**Comment by inklesspen** [comment]
@atesh Anything more I can do to help here? I don't know a lot about the internals of xkit, unfortunately.

**Comment by inklesspen** [comment]
It occurs to me there's no inherent reason we couldn't implement our own post editor.

**Comment by dotWee** [comment]
I'm curious, am I right if I say, tumblr is encrypting everything which gets typed using the new editor?
